src/wp-includes/global-styles-and-settings.php

```diff
@@ -407,8 +407,23 @@ function wp_theme_has_theme_json() {
 		return $theme_has_support;
 	}
 
-	// Does the theme have its own theme.json?
-	$theme_has_support = is_readable( get_theme_file_path( 'theme.json' ) );
+	$stylesheet_directory = get_stylesheet_directory();
+	$template_directory   = get_template_directory();
+	$directories          = array( $stylesheet_directory );
+	if ( $stylesheet_directory !== $template_directory ) {
+		$directories[] = $template_directory;
+	}
+
+	$theme_has_support = false;
+	foreach ( $directories as $directory ) {
+		/** This filter is documented in wp-includes/link-template.php */
+		$file_path = apply_filters( 'theme_file_path', $directory . '/theme.json', 'theme.json' );
+		// Does the theme have its own theme.json?
+		$theme_has_support = is_readable( $file_path );
+		if ( $theme_has_support ) {
+			break;
+		}
+	}
 
 	return $theme_has_support;
```

Comment:
With the above in mind it may be easier to use the same execution path as the function does.

```suggestion
	if ( file_exists( get_stylesheet_directory() . '/theme.json' ) ) {
		$path = get_stylesheet_directory() . '/theme.json';
	} else {
		$path = get_template_directory() . '/theme.json';
	}
	
	/** This filter is documented in wp-includes/link-template.php */
	$path = apply_filters( 'theme_file_path', $path, 'theme.json' );
	
	if ( file_exists( $path ) ) {
		$theme_has_support = true;
	} else {
		$theme_has_support = false;
	}
	
	return $theme_has_support;
```

Reply:
FWIW This is much easier for me to read, rather than a loop that will run 1-2 times.

I would document that it's the same as `get_theme_file_path( 'theme.json' )` but that's not loaded in load-styles.php context.

Reply:
Self merging above having discussed with others. This will allow WP-Dev to unblock GB.